### PR TITLE
[PAB-318] Ingest data from Towns Fund spreadsheet

### PR DESCRIPTION
**Pull request by nsnape**
PAB-318: https://trello.com/c/yCVpmSB1 

### Change description
Methods to extract data from TF data entry form (Excel file).

Each sheet needs it's own method due to large differences between sheets. Each extracted added to it's own dataframe, which are attached to a dict. 

- [ ] Unit tests and other appropriate tests added or updated
- [ ] README and other documentation has been updated / added (if needed)
- [ ] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
_If manual testing is needed, give suggested testing steps_

### CODE REVIEW

File: core/extraction/towns_fund.py [hunk 1]
```diff
@@ -0,0 +1,454 @@
+"""
+Methods specifically for extracting data from Towns Fund reporting template (Excel Spreadsheet)
+"""
+import numpy as np
+import pandas as pd
+
+
+def ingest_towns_fund_data(df_ingest: pd.DataFrame) -> dict[pd.DataFrame]:
```
Comment (cyrusdobbs) [inline core/extraction/towns_fund.py:8]: return hint should be `dict[str, pd.DataFrame]`

File: core/extraction/towns_fund.py [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+        df_ingest["4a - Funding Profiles"], number_of_projects
+    )
+    towns_fund_extracted["df_funding_extracted"] = extract_funding_data(
+        df_ingest["4a - Funding Profiles"], number_of_projects
+    )
+
+    towns_fund_extracted["df_psi_extracted"] = extract_psi(df_ingest["4b - PSI"])
+
+    towns_fund_extracted["df_outputs_extracted"] = extract_outputs(df_ingest["5 - Project Outputs"], number_of_projects)
+    towns_fund_extracted["df_outcomes_extracted"] = extract_outcomes(df_ingest["6 - Outcomes"])
```
Comment (cyrusdobbs) [inline core/extraction/towns_fund.py:39]: IDE is complaining saying these could all be put into one big dictionary instantiation but I guess that's not quite as easy to debug?..

File: core/extraction/towns_fund.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+    """
+
+    # strip out everything other than "SECTION A..." in spreadsheet.
+    df_programme = df_programme.iloc[5:20, 2:5]
+
+    # rename col headers for ease
+    df_programme.columns = [0, 1, 2]
+
+    # combine first 2 cols into 1, filling in blank (merged) cells
+    field_names_first = [x := y if y is not np.nan else x for y in df_programme[0]]  # noqa: F841,F821
```
Comment (cyrusdobbs) [inline core/extraction/towns_fund.py:71]: How come you're using the `:=` notation here? I can't see how it's being used but I might be missing something?

File: core/extraction/towns_fund.py (hunk 2 above)
Comment (nsnape) [reply core/extraction/towns_fund.py:39]: yeah I thought about thta, but decided to leave it verbosely line-by-line, for the purpose of debugging, as you say

File: core/extraction/towns_fund.py (hunk 3 above)
Comment (cyrusdobbs) [reply core/extraction/towns_fund.py:71]: ahh very snazzy!

**Review by cyrusdobbs — APPROVED** [review]

File: core/extraction/towns_fund.py (hunk 3 above)
Comment (nsnape) [reply core/extraction/towns_fund.py]: the only minor downer on this is the x also persists in namespace after the comprehension has finished... :D

**Pull request merged**